core/src/storage2/collections/binary_heap/wrapper.rs

```diff
@@ -0,0 +1,284 @@
+//! elements of the [`BinaryHeap`](`super::BinaryHeap`) in storage.
+
+use crate::storage2::{
+    collections::binary_heap::{
+        group,
+        Group,
+        Iter,
+        IterMut,
+        StorageVec,
+    },
+    traits::{
+        KeyPtr,
+        PackedLayout,
+        SpreadLayout,
+    },
+    Lazy,
+};
+
+/// The wrapper provides an interface for accessing elements.
+/// Since elements are stored in groups of two the requested element
+/// index needs to be transposed to the group in which the element
+/// is stored.
+#[derive(Default, PartialEq, Eq, Debug)]
+pub struct Wrapper<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// The number of elements stored in the heap.
+    /// We cannot use the length of the storage vector, since each entry (= `Group`)
+    /// in the vector contains two child elements (except the root element which occupies
+    /// a `Group` on its own.
+    len: Lazy<u32>,
+    /// The underlying storage vec containing the grouped elements.
+    elems: StorageVec<Group<T>>,
+}
+
+impl<T> Wrapper<T>
+where
+    T: PackedLayout + Ord,
+{
+    /// Creates a new empty storage heap.
+    pub fn new() -> Self {
+        Self {
+            len: Lazy::new(0),
+            elems: StorageVec::new(),
+        }
+    }
+
+    /// Returns the number of elements in the heap, also referred to as its 'length'.
+    pub fn len(&self) -> u32 {
+        *self.len
+    }
+
+    /// Returns the amount of groups stored in the vector.
+    #[allow(dead_code)]
+    #[cfg(all(test, feature = "ink-fuzz-tests"))]
+    pub fn group_count(&self) -> u32 {
+        self.elems.len()
+    }
+
+    /// Returns `true` if the heap contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns a shared reference to the indexed element.
+    ///
+    /// Returns `None` if `index` is out of bounds.
+    pub fn get(&self, index: u32) -> Option<&T> {
+        let group_index = group::get_group_index(index);
+        let i = self.within_bounds(group_index)?;
+        let group = self.elems.get(i)?;
+        group.as_ref(index)
+    }
+
+    /// Returns an exclusive reference to the indexed element.
+    /// The element in a group is an `Option<T>`.
+    ///
+    /// Returns `None` if `index` is out of bounds.
+    pub fn get_mut(&mut self, index: u32) -> Option<&mut Option<T>> {
+        let group_index = group::get_group_index(index);
+        self.within_bounds(group_index)?;
+        self.elems
+            .get_mut(group_index)
+            .map(|group| group.get_mut(index))
+    }
+
+    /// Swaps the elements at the given indices.
+    ///
+    /// # Panics
+    ///
+    /// If one or both indices are out of bounds.
+    pub fn swap(&mut self, a: u32, b: u32) {
+        if a == b {
+            return
+        }
+        assert!(
+            a < self.len() && b < self.len(),
+            "indices are out of bounds"
+        );
```

Comment:
Wondering whether it would be better like:

https://github.com/paritytech/ink/blob/59dcb1bad0250d581217a5c22af1f9ac3071a5b4/core/src/storage2/lazy/lazy_array.rs#L562-L563